File: Java/src/test/java/workshops/testingil/dirtytests/exercise/tests1.java
package workshops.testingil.dirtytests.exercise;

import io.github.bonigarcia.wdm.WebDriverManager;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.params.ParameterizedTest;
import org.junit.jupiter.params.provider.CsvFileSource;
import org.openqa.selenium.By;
import org.openqa.selenium.WebDriver;
import org.openqa.selenium.WebElement;
import org.openqa.selenium.chrome.ChromeDriver;
import org.openqa.selenium.chrome.ChromeOptions;
import org.springframework.web.reactive.function.client.WebClient;
import workshops.testingil.dirtytests.app.Calculator;
import workshops.testingil.dirtytests.app.ResultWrapper;

import java.net.URLEncoder;
import java.nio.charset.StandardCharsets;
import java.util.concurrent.TimeUnit;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertTrue;

public class tests1 {

    private WebDriver driver;
    public static final String URL = "C:\\GitHub\\Presentations\\Dirty-Tests-Workshop\\Java\\src\\UI\\CalculatorUI.html";


    @Test
    public void test_1_2_3()
    {
        Calculator c = new Calculator();
        c.press("1");
        c.press("2");
        c.press("3");
        assertEquals(c.getDisplay(),"123");
    }


    @Test
//@Disabled
    public void test_plus_and_2() {

        WebClient client = WebClient.create("http://localhost:8080");
        ResultWrapper result;

        client.post()
                .uri(uriBuilder -> uriBuilder
                        .path("/calc/reset")
                        .build())
                .retrieve().bodyToMono(Void.class).block();

        client.post()
                .uri(uriBuilder -> uriBuilder
                        .path("/calc/press")
                        .queryParam("key", URLEncoder.encode("+", StandardCharsets.UTF_8))
                        .build())
                .retrieve().bodyToMono(Void.class).block();


        client.post()
                .uri(uriBuilder -> uriBuilder
                        .path("/calc/press")
                        .queryParam("key", "2")
                        .build())
                .retrieve().bodyToMono(Void.class).block();

        result = client.get()
                .uri(uriBuilder -> uriBuilder
                        .path("/calc/display")
                        .build())
                .retrieve().bodyToMono(ResultWrapper.class).block();

        assertEquals("2", result.display);

    }

    @Test
    public void test_ui_1_plus_34() {
        WebDriverManager.chromedriver().setup();
        var options = new ChromeOptions();
        options.addArguments("remote-allow-origins=*");
        driver = new ChromeDriver(options);
        driver.manage().timeouts().implicitlyWait(10, TimeUnit.SECONDS);

        driver.get(URL);
        String BUTTON_XPATH4 = "//*[@id=\"id" + "reset" + "\"]";
        WebElement keyElement4 = driver.findElement(By.xpath(BUTTON_XPATH4));
        keyElement4.click();
        try {
            Thread.sleep(500);
        } catch (InterruptedException e4) {
            System.out.println("Timeout!!!");
        }
        String BUTTON_XPATH3 = "//*[@id=\"id" + "1" + "\"]";
        WebElement keyElement3 = driver.findElement(By.xpath(BUTTON_XPATH3));
        keyElement3.click();
        try {
            Thread.sleep(500);
        } catch (InterruptedException e3) {
            System.out.println("Timeout!!!");
        }
        String BUTTON_XPATH2 = "//*[@id=\"id" + "plus" + "\"]";
        WebElement keyElement2 = driver.findElement(By.xpath(BUTTON_XPATH2));
        keyElement2.click();
        try {
            Thread.sleep(500);
        } catch (InterruptedException e2) {
            System.out.println("Timeout!!!");
        }
        String BUTTON_XPATH1 = "//*[@id=\"id" + "3" + "\"]";
        WebElement keyElement1 = driver.findElement(By.xpath(BUTTON_XPATH1));
        keyElement1.click();
        try {
            Thread.sleep(500);
        } catch (InterruptedException e1) {
            System.out.println("Timeout!!!");
        }
        String BUTTON_XPATH = "//*[@id=\"id" + "4" + "\"]";
        WebElement keyElement = driver.findElement(By.xpath(BUTTON_XPATH));
        keyElement.click();
        try {
            Thread.sleep(500);
        } catch (InterruptedException e) {
            System.out.println("Timeout!!!");
        }
        String BUTTON_XPATH5 = "//*[@id=\"id" + "Update" + "\"]";
        WebElement keyElement5 = driver.findElement(By.xpath(BUTTON_XPATH5));
        keyElement5.click();
        try {
            Thread.sleep(500);
        } catch (InterruptedException e) {
            System.out.println("Timeout!!!");
        }

        String DISPLAY_XPATH = "//*[@id=\"result\"]";
        WebElement res = driver.findElement(By.xpath(DISPLAY_XPATH));
        String result = res.getAttribute("value");
        assertEquals("34", result);
        driver.quit();
    }

    @ParameterizedTest
    @CsvFileSource(resources = "/slow_unit_test_data.csv")
    public void test_multiple_vals(String in, String out){
        Calculator c = new Calculator();
        in.chars()
                .forEach(ch -> c.press(String.valueOf((char)ch)));
        assertEquals(c.getDisplay(), out);
    }

    @Test
    public void test_nothing(){
        Calculator c = new Calculator();
        assertTrue(c.getDisplay() == "0");
    }


    @Test
    //@Disabled
    public void test_1_and_plus() {

        WebClient client = WebClient.create("http://localhost:8080");
        ResultWrapper result;

        client.post()
                .uri(uriBuilder -> uriBuilder
                        .path("/calc/reset")
                        .build())
                .retrieve().bodyToMono(Void.class).block();

        client.post()
                .uri(uriBuilder -> uriBuilder
                        .path("/calc/press")
                        .queryParam("key", "1")
                        .build())
                .retrieve().bodyToMono(Void.class).block();


        client.post()
                .uri(uriBuilder -> uriBuilder
                        .path("/calc/press")
                        .queryParam("key", URLEncoder.encode("+", StandardCharsets.UTF_8))
                        .build())
                .retrieve().bodyToMono(Void.class).block();

        result = client.get()
                .uri(uriBuilder -> uriBuilder
                        .path("/calc/display")
                        .build())
                .retrieve().bodyToMono(ResultWrapper.class).block();

        assertEquals("1", result.display);

    }

    @Test
    public void test_1(){
        Calculator c = new Calculator();
        c.press("1");
        assertTrue(c.getDisplay() == "1");
    }

    @Test
    public void test_zero_zero() {

        WebClient client = WebClient.create("http://localhost:8080");
        ResultWrapper result;

        client.post()
                .uri(uriBuilder -> uriBuilder
                        .path("/calc/reset")
                        .build())
                .retrieve().bodyToMono(Void.class).block();

        client.post()
                .uri(uriBuilder -> uriBuilder
                        .path("/calc/press")
                        .queryParam("key", "0")
                        .build())
                .retrieve().bodyToMono(Void.class).block();

        client.post()
                .uri(uriBuilder -> uriBuilder
                        .path("/calc/press")
                        .queryParam("key", "0")
                        .build())
                .retrieve().bodyToMono(Void.class).block();


        result = client.get()
                .uri(uriBuilder -> uriBuilder
                        .path("/calc/display")
                        .build())
                .retrieve().bodyToMono(ResultWrapper.class).block();

        assertEquals("0", result.display);

    }

    @Test
    public void test_div_6_2() {
        WebDriverManager.chromedriver().setup();
        var options = new ChromeOptions();
        options.addArguments("remote-allow-origins=*");
        driver = new ChromeDriver(options);
        driver.manage().timeouts().implicitlyWait(10, TimeUnit.SECONDS);
        driver.get(URL);
        String BUTTON_XPATH4 = "//*[@id=\"id" + "reset" + "\"]";
        WebElement keyElement4 = driver.findElement(By.xpath(BUTTON_XPATH4));
        keyElement4.click();
        try {
            Thread.sleep(500);
        } catch (InterruptedException e4) {
            System.out.println("Timeout!!!");
        }
        String BUTTON_XPATH3 = "//*[@id=\"id" + "6" + "\"]";
        WebElement keyElement3 = driver.findElement(By.xpath(BUTTON_XPATH3));
        keyElement3.click();
        try {
            Thread.sleep(500);
        } catch (InterruptedException e3) {
            System.out.println("Timeout!!!");
        }
        String BUTTON_XPATH2 = "//*[@id=\"id" + "div" + "\"]";
        WebElement keyElement2 = driver.findElement(By.xpath(BUTTON_XPATH2));
        keyElement2.click();
        try {
            Thread.sleep(500);
        } catch (InterruptedException e2) {
            System.out.println("Timeout!!!");
        }
        String BUTTON_XPATH1 = "//*[@id=\"id" + "2" + "\"]";
        WebElement keyElement1 = driver.findElement(By.xpath(BUTTON_XPATH1));
        keyElement1.click();
        try {
            Thread.sleep(500);
        } catch (InterruptedException e1) {
            System.out.println("Timeout!!!");
        }
        String BUTTON_XPATH = "//*[@id=\"id" + "eql" + "\"]";
        WebElement keyElement = driver.findElement(By.xpath(BUTTON_XPATH));
        keyElement.click();
        try {
            Thread.sleep(500);
        } catch (InterruptedException e) {
            System.out.println("Timeout!!!");
        }
        String BUTTON_XPATH5 = "//*[@id=\"id" + "Update" + "\"]";
        WebElement keyElement5 = driver.findElement(By.xpath(BUTTON_XPATH5));
        keyElement5.click();
        try {
            Thread.sleep(500);
        } catch (InterruptedException e) {
            System.out.println("Timeout!!!");
        }

        String DISPLAY_XPATH = "//*[@id=\"result\"]";
        WebElement res = driver.findElement(By.xpath(DISPLAY_XPATH));
        String result = res.getAttribute("value");
        assertEquals("3", result);
        driver.quit();
    }


    @Test
    public void test_1_and_C()
    {
        Calculator c = new Calculator();
        c.press("1");
        c.press("C");
        assertEquals(c.getDisplay(),"0");
    }

    @Test
    public void test_0() {
        WebClient client = WebClient.create("http://localhost:8080");
        ResultWrapper result;

        client.post()
                .uri(uriBuilder -> uriBuilder
                        .path("/calc/press")
                        .queryParam("key", "0")
                        .build())
                .retrieve().bodyToMono(Void.class).block();


        result = client.get()
                .uri(uriBuilder -> uriBuilder
                        .path("/calc/display")
                        .build())
                .retrieve().bodyToMono(ResultWrapper.class).block();

        assertEquals("0", result.display);
    }

}

File: CSharp/DirtyTestsWorkshop/Exercise/Tests1.cs
using CsvHelper;
using DirtyTestsWorkshop.App;
using Newtonsoft.Json;
using NUnit.Framework;
using OpenQA.Selenium;
using OpenQA.Selenium.Chrome;
using System.Globalization;
using System.Web;

namespace DirtyTestsWorkshop.Exercise
{
    [TestFixture]
    internal class Tests1
    {
        private WebDriver driver;
        private const string path_to_file = @"Exercise\Resources\slow_unit_test_data.csv";

        [Test]
        public void test_1_2_3()
        {
            Calculator c = new Calculator();
            c.Press("1");
            c.Press("2");
            c.Press("3");
            Assert.AreEqual(c.GetDisplay(), "123");
        }

        [Test]
        //[Ignore]
        public async Task test_plus_and_2()
        {
            HttpClient client = new HttpClient();
            ResultWrapper result;

            await client.PostAsync("http://localhost:8080/calc/reset", null);

            var builder = new UriBuilder("http://localhost:8080/calc/press");
            builder.Query = "key=" + HttpUtility.UrlEncode("+");
            await client.PostAsync(builder.ToString(), null);

            builder.Query = "key=2";
            await client.PostAsync(builder.ToString(), null);

            HttpResponseMessage response = await client.GetAsync("http://localhost:8080/calc/display");
            var body = await response.Content.ReadAsStringAsync();
            result = JsonConvert.DeserializeObject<ResultWrapper>(body);
            Assert.AreEqual("2", result.display);

        }
        [Test]
        public void test_ui_1_plus_34()
        {
            var URL = "file:///" + Environment.CurrentDirectory + "/CalculatorUI.html";
            ChromeOptions options = new ChromeOptions();
            options.AddArguments("remote-allow-origins=*");
            driver = new ChromeDriver(options);
            driver.Manage().Timeouts().ImplicitWait = TimeSpan.FromSeconds(10);
            driver.Url = URL;

            String BUTTON_XPATH4 = "//*[@id=\'id" + "reset" + "\']";
            IWebElement keyElement4 = driver.FindElement(By.XPath(BUTTON_XPATH4));
            keyElement4.Click();

            Thread.Sleep(500);

            String BUTTON_XPATH3 = "//*[@id=\'id" + "1" + "\']";
            IWebElement keyElement3 = driver.FindElement(By.XPath(BUTTON_XPATH3));
            keyElement3.Click();
            Thread.Sleep(500);

            String BUTTON_XPATH2 = "//*[@id=\'id" + "plus" + "\']";
            IWebElement keyElement2 = driver.FindElement(By.XPath(BUTTON_XPATH2));
            keyElement2.Click();
            Thread.Sleep(500);

            String BUTTON_XPATH1 = "//*[@id=\'id" + "3" + "\']";
            IWebElement keyElement1 = driver.FindElement(By.XPath(BUTTON_XPATH1));
            keyElement1.Click();
            Thread.Sleep(500);

            String BUTTON_XPATH = "//*[@id=\'id" + "4" + "\']";
            IWebElement keyElement = driver.FindElement(By.XPath(BUTTON_XPATH));
            keyElement.Click();
            Thread.Sleep(500);

            String BUTTON_XPATH5 = "//*[@id=\'id" + "Update" + "\']";
            IWebElement keyElement5 = driver.FindElement(By.XPath(BUTTON_XPATH5));
            keyElement5.Click();
            Thread.Sleep(500);

            String DISPLAY_XPATH = "//*[@id=\"result\"]";
            IWebElement res = driver.FindElement(By.XPath(DISPLAY_XPATH));
            String result = res.GetAttribute("value");
            Assert.AreEqual("34", result);

            driver.Quit();
        }

        [Test, TestCaseSource("GetTestData")]
        public void test_multiple_vals(string input, string output)
        {
            Calculator c = new Calculator();
            foreach (char ch in input)
            {
                c.Press(ch.ToString());
            }
            Assert.AreEqual(c.GetDisplay(), output);
        }

        private static IEnumerable<string[]> GetTestData()
        {
            using (var reader = new StreamReader(path_to_file))
            using (var csv = new CsvReader(reader, CultureInfo.InvariantCulture))
            {
                while (csv.Read())
                {
                    yield return new string[] { csv[0], csv[1] };

                }
            }
        }

        [Test]
        public void test_nothing()
        {
            Calculator c = new Calculator();
            Assert.IsTrue(c.GetDisplay() == "0");
        }

        [Test]
        //[Ignore]
        public async Task test_1_and_plus()
        {

            HttpClient client = new HttpClient();
            ResultWrapper result;

            await client.PostAsync("http://localhost:8080/calc/reset", null);

            var builder = new UriBuilder("http://localhost:8080/calc/press");
            builder.Query = "key=1";

            await client.PostAsync(builder.ToString(), null);

            builder.Query = "key=" + HttpUtility.UrlEncode("+");
            await client.PostAsync(builder.ToString(), null);

            HttpResponseMessage response = await client.GetAsync("http://localhost:8080/calc/display");
            var body = await response.Content.ReadAsStringAsync();
            result = JsonConvert.DeserializeObject<ResultWrapper>(body);
            Assert.AreEqual("1", result.display);

        }

        [Test]
        public void test_1()
        {
            Calculator c = new Calculator();
            c.Press("1");
            Assert.IsTrue(c.GetDisplay() == "1");
        }

        [Test]
        public async Task test_zero_zero()
        {
            HttpClient client = new HttpClient();
            ResultWrapper result;

            await client.PostAsync("http://localhost:8080/calc/reset", null);

            var builder = new UriBuilder("http://localhost:8080/calc/press");
            builder.Query = "key=0";

            await client.PostAsync(builder.ToString(), null);
            await client.PostAsync(builder.ToString(), null);

            HttpResponseMessage response = await client.GetAsync("http://localhost:8080/calc/display");
            var body = await response.Content.ReadAsStringAsync();
            result = JsonConvert.DeserializeObject<ResultWrapper>(body);
            Assert.AreEqual("0", result.display);
        }

        [Test]
        public void test_div_6_2()
        {
            var URL = "file:///" + Environment.CurrentDirectory + "/CalculatorUI.html";
            ChromeOptions options = new ChromeOptions();
            options.AddArguments("remote-allow-origins=*");
            driver = new ChromeDriver(options);
            driver.Manage().Timeouts().ImplicitWait = TimeSpan.FromSeconds(10);
            driver.Url = URL;

            String BUTTON_XPATH4 = "//*[@id=\'id" + "reset" + "\']";
            IWebElement keyElement4 = driver.FindElement(By.XPath(BUTTON_XPATH4));
            keyElement4.Click();
            Thread.Sleep(500);

            String BUTTON_XPATH3 = "//*[@id=\'id" + "6" + "\']";
            IWebElement keyElement3 = driver.FindElement(By.XPath(BUTTON_XPATH3));
            keyElement3.Click();
            Thread.Sleep(500);

            String BUTTON_XPATH2 = "//*[@id=\'id" + "div" + "\']";
            IWebElement keyElement2 = driver.FindElement(By.XPath(BUTTON_XPATH2));
            keyElement2.Click();
            Thread.Sleep(500);

            String BUTTON_XPATH1 = "//*[@id=\'id" + "2" + "\']";
            IWebElement keyElement1 = driver.FindElement(By.XPath(BUTTON_XPATH1));
            keyElement1.Click();
            Thread.Sleep(500);

            String BUTTON_XPATH = "//*[@id=\'id" + "eql" + "\']";
            IWebElement keyElement = driver.FindElement(By.XPath(BUTTON_XPATH));
            keyElement.Click();
            Thread.Sleep(500);

            String BUTTON_XPATH5 = "//*[@id=\'id" + "Update" + "\']";
            IWebElement keyElement5 = driver.FindElement(By.XPath(BUTTON_XPATH5));
            keyElement5.Click();
            Thread.Sleep(500);

            String DISPLAY_XPATH = "//*[@id=\'result\']";
            IWebElement res = driver.FindElement(By.XPath(DISPLAY_XPATH));
            String result = res.GetAttribute("value");
            Assert.AreEqual("3", result);
            driver.Quit();
        }

        [Test]
        public void test_1_and_C()
        {
            Calculator c = new Calculator();
            c.Press("1");
            c.Press("C");
            Assert.AreEqual(c.GetDisplay(), "0");
        }

        [Test]
        public async Task test_0()
        {
            HttpClient client = new HttpClient();

            var builder = new UriBuilder("http://localhost:8080/calc/press");
            builder.Query = "key=0";
            var url = builder.ToString();

            await client.PostAsync(url, null);

            HttpResponseMessage response = await client.GetAsync("http://localhost:8080/calc/display");
            var body = await response.Content.ReadAsStringAsync();
            ResultWrapper result = JsonConvert.DeserializeObject<ResultWrapper>(body);
            Assert.AreEqual("0", result.display);
        }


    }
}
